joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h

```diff
@@ -348,145 +348,20 @@ template <class SegmentImpl, class HardwareInterface>
-            checkStateTolerancePerJoint(state_joint_error_, joint_tolerances.state_tolerance, true);
-          }
-          rt_segment_goal->preallocated_result_->error_code =
-          control_msgs::FollowJointTrajectoryResult::PATH_TOLERANCE_VIOLATED;
-          rt_segment_goal->setAborted(rt_segment_goal->preallocated_result_);
-          rt_active_goal_.reset();
-          successful_joint_traj_.reset();
-        }
-      }
-      else if (segment_it == --curr_traj[i].end())
-      {
-        if (verbose_)
-          ROS_DEBUG_STREAM_THROTTLE_NAMED(1,name_,"Finished executing last segment, checking goal tolerances");
-
-        // Controller uptime
-        const ros::Time uptime = time_data_.readFromRT()->uptime;
-
-        // Checks that we have ended inside the goal tolerances
-        const SegmentTolerancesPerJoint<Scalar>& tolerances = segment_it->getTolerances();
-        const bool inside_goal_tolerances = checkStateTolerancePerJoint(state_joint_error_, tolerances.goal_state_tolerance);
-
-        if (inside_goal_tolerances)
-        {
-          successful_joint_traj_[i] = 1;
-        }
-        else if (uptime.toSec() < segment_it->endTime() + tolerances.goal_time_tolerance)
-        {
-          // Still have some time left to meet the goal state tolerances
-        }
-        else
-        {
-          if (verbose_)
-          {
-            ROS_ERROR_STREAM_NAMED(name_,"Goal tolerances failed for joint: "<< joint_names_[i]);
-            // Check the tolerances one more time to output the errors that occurs
-            checkStateTolerancePerJoint(state_joint_error_, tolerances.goal_state_tolerance, true);
-          }
+  updateStateAndTimeData(time, period);
 
-          rt_segment_goal->preallocated_result_->error_code = control_msgs::FollowJointTrajectoryResult::GOAL_TOLERANCE_VIOLATED;
-          rt_segment_goal->setAborted(rt_segment_goal->preallocated_result_);
-          rt_active_goal_.reset();
-          successful_joint_traj_.reset();
-        }
-      }
-    }
-  }
-
-  //If there is an active goal and all segments finished successfully then set goal as succeeded
-  RealtimeGoalHandlePtr current_active_goal(rt_active_goal_);
-  if (current_active_goal && successful_joint_traj_.count() == joints_.size())
+  if (allSegmentsCompleted())
   {
-    current_active_goal->preallocated_result_->error_code = control_msgs::FollowJointTrajectoryResult::SUCCESSFUL;
-    current_active_goal->setSucceeded(current_active_goal->preallocated_result_);
-    current_active_goal.reset(); // do not publish feedback
-    rt_active_goal_.reset();
-    successful_joint_traj_.reset();
+    setGoalAsSucceeded();
   }
 
   // Hardware interface adapter: Generate and send commands
-  hw_iface_adapter_.updateCommand(time_data.uptime, time_data.period,
+  hw_iface_adapter_.updateCommand(time_data_.readFromRT()->uptime, time_data_.readFromRT()->period,
```

Comment:
Ok, reading period from `time_data_` is unnecessary.

I can not keep the timedata caching at the beginning since I would have to break up the structure of the new update-function:

>   // NOTE: It is very important to execute the two above code blocks in the specified sequence: first get current
  // trajectory, then update time data.

My latest approach is passing time_data by reference. What do you think?